### PR TITLE
(core) Support UnitAttr in assembly format

### PR DESCRIPTION
**Pull request by n-io**
This PR enables support for UnitAttr in the assembly format parser.

The upstream MLIR assembly format syntax to represent a UnitAttr may be for instance ([source](https://mlir.llvm.org/docs/Dialects/BufferizationOps/#bufferizationto_memref-bufferizationtomemrefop)):
```
`bufferization.to_memref` $tensor (`read_only` $read_only^)? attr-dict `:` type($memref)
```

Specifically, if the keyword `read_only` is present, then the UnitAttr (of the same name) will be set. In xdsl, it would simply be set by assigning `= UnitAttr()`. The approach here is similar, if the optional keyword is present, then the value should be set.

This change consists of two parts:
* The first is to not change the `unique_base` variable in the assembly format parser in the way it's usually done for builtins. This causes `AttributeVariable.parse` to not execute the `parser.parse_attribute()` in the `unique_base is None` case (expecting to parse an actual value), but instead to use the custom parsing method of UnitAttr in the `issubclass(.. , ParametrizedAttribute)` case
* The second is to implement a `UnitAttr.parse_parameters` method returning an empty list, overriding the parse method of the superclass which attempts to parse an actual value.

### CODE REVIEW

File: xdsl/irdl/declarative_assembly_format.py [hunk 1]
```diff
@@ -11,6 +11,7 @@
 from dataclasses import dataclass, field
 from typing import Any, Literal, cast
 
+from xdsl.dialects.builtin import UnitAttr
```
Comment (superlopuh) [inline xdsl/irdl/declarative_assembly_format.py:14]: this seems unfortunate, I would like to avoid importing any dialects here, it seems like it breaks the layering conceptually

Comment (n-io) [reply xdsl/irdl/declarative_assembly_format.py:14]: Can we live with the fact that it would print an extra whitespace?

Comment (superlopuh) [reply xdsl/irdl/declarative_assembly_format.py:14]: Where is the whitespace printed?

File: xdsl/dialects/builtin.py [hunk 2]
```diff
@@ -431,6 +431,11 @@ def bitwidth(self) -> int:
 class UnitAttr(ParametrizedAttribute):
     name = "unit"
 
+    @classmethod
+    def parse_parameters(cls, parser: AttrParser) -> Sequence[Attribute]:
+        """Parse the attribute parameters."""
+        return []
+
```
Comment (superlopuh) [inline xdsl/dialects/builtin.py:438]: why is this necessary?

File: tests/irdl/test_declarative_assembly_format.py [hunk 3]
```diff
@@ -455,6 +455,68 @@ class OptionalPropertyOp(IRDLOperation):
...
+)
+def test_optional_unit_attr_attribute(program: str, generic_program: str):
+    """Test the parsing of optional UnitAttr operands"""
+
+    @irdl_op_definition
+    class OptionalUnitAttrOp(IRDLOperation):
+        name = "test.optional_unit_attr"
+        unit_attr = opt_prop_def(UnitAttr)
+
+        assembly_format = "(`unit_attr` $unit_attr^)? attr-dict"
```
Comment (superlopuh) [inline tests/irdl/test_declarative_assembly_format.py:510]: Could you please make the strings different to disambiguate? Maybe something like this?

```suggestion
        unit_attr_prop = opt_prop_def(UnitAttr)

        assembly_format = "(`unit_attr_str` $unit_attr_prop^)? attr-dict"
```

File: xdsl/irdl/declarative_assembly_format.py (hunk 1 above)
Comment (PapyChacal) [reply xdsl/irdl/declarative_assembly_format.py:14]: > this seems unfortunate, I would like to avoid importing any dialects here, it seems like it breaks the layering conceptually

I don't think we could avoid it *and* give special treatment to those here, which is necessary for compatibility

Comment (superlopuh) [reply xdsl/irdl/declarative_assembly_format.py]: But I don't understand the special treatment. Could you maybe post what happens today without these changes?

Comment (superlopuh) [reply xdsl/irdl/declarative_assembly_format.py]: Ah I see it was added in the latest commit. Which is the line of code that prints the extra space? Shouldn't it be avoided? Do we want this to work differently for UnitAttr to any other parametrized attribute with no parameters?

Comment (PapyChacal) [reply xdsl/irdl/declarative_assembly_format.py]: I think the support for UnitAttr should be structured differently too; I think a custom directive would make more sense, rather than adapting the current one to special case UnitAttr.

Comment (PapyChacal) [reply xdsl/irdl/declarative_assembly_format.py]: More specifically, I think that rather than hacking UnitAttr, the generic name-eliding feature and all to behave syntactically like we want, it would be way clearer to have a custom directive (maybe an OptionaVariable subclass) encoding the corner case of UnitAttr.
UnitAttr in declarative syntax is a corner-case by itself, and I think it would be clearer this way.
https://mlir.llvm.org/docs/DefiningDialects/Operations/#unit-attributes

File: xdsl/dialects/builtin.py (hunk 2 above)
Comment (PapyChacal) [reply xdsl/dialects/builtin.py:438]: See my comment on the other thread on this
https://github.com/xdslproject/xdsl/pull/2995#discussion_r1709570381

File: xdsl/irdl/declarative_assembly_format.py (hunk 1 above)
Comment (PapyChacal) [reply xdsl/irdl/declarative_assembly_format.py]: I'm happy to have a short call to clarify if it's unclear - I know that part of the code base can be a bit overwhelming :upside_down_face:

File: tests/irdl/test_declarative_assembly_format.py (hunk 3 above)
Comment (n-io) [reply tests/irdl/test_declarative_assembly_format.py:510]: Before doing this, I'd note that this would diverge from how the other tests do it, as for the ones immediately above and below

Comment (superlopuh) [reply tests/irdl/test_declarative_assembly_format.py:510]: OK fair enough, please disregard

**Comment by superlopuh** [comment]
OK, thank you both for the explanations and discussion. I see why the special handling is needed, indeed there seems to not be a much better way to do this than with special handling in the printer and parser. I suggest reverting to the state before my suggestions and continuing from there?

File: xdsl/irdl/declarative_assembly_format.py (hunk 1 above)
Comment (superlopuh) [reply xdsl/irdl/declarative_assembly_format.py]: No time for calls today, let's just keep iterating here

Comment (PapyChacal) [reply xdsl/irdl/declarative_assembly_format.py:14]: Change of direction done!

File: xdsl/irdl/declarative_assembly_format.py [hunk 4]
```diff
@@ -801,6 +802,21 @@ class OptionalAttributeVariable(AttributeVariable, OptionalVariable):
     """
 
 
+class OptionalUnitAttrVariable(OptionalAttributeVariable):
+    """
+    An optional UnitAttr variable
```
Comment (superlopuh) [inline xdsl/irdl/declarative_assembly_format.py:807]: Could you please expand a bit here why this is needed? Something about UnitAttr not being printed in the IR, so special handling is needed, maybe with a link to the MLIR documentation you posted in the other comment.

**Review by superlopuh — APPROVED** [review]
Thank you!

**Review by PapyChacal — APPROVED** [review]

File: xdsl/irdl/declarative_assembly_format.py (hunk 4 above)
Comment (n-io) [reply xdsl/irdl/declarative_assembly_format.py]: Doing as suggested
